### PR TITLE
Fixes SI-6556 by converting asserts to logging.

### PR DESCRIPTION
**Pull request by odersky**
Previous fix to value classes uncovered some questionable cases in the backend where result types of constructor signatures are surprising. It's not a big deal because these types will be ignored afterwards anyway. But
the method uncovered some questionable situations which we should follow up on. However, breaking 2.9 code because of this is way too harsh. That's why the asserts were converted to warnings.

### CODE REVIEW

**Comment by odersky** [comment]
Review by @adriaanm, @paulp

**Comment by paulp** [comment]
Seems to be missing the test case (I think the one at https://issues.scala-lang.org/secure/attachment/15306/Series.scala is ready-made.)

And as long as it needs that change anyway, our logging output would be a lot more useful if log messages which meant "surely wrong thing is happening" were marked differently from "here's what's happening." While trying to get around to a unified logging situation, I have been prefixing the log message with !!! like log("!!! something is happening"). This also serves to mark the source code.

File: src/reflect/scala/reflect/internal/transform/Erasure.scala [hunk 1]
```diff
@@ -214,14 +214,29 @@ trait Erasure {
...
       case ExistentialType(tparams, restpe) =>
         specialConstructorErasure(clazz, restpe)
+      case RefinedType(parents, decls) =>
+        specialConstructorErasure(
+          clazz, specialScalaErasure.mergeParents(parents))
       case mt @ MethodType(params, restpe) =>
         MethodType(
           cloneSymbolsAndModify(params, specialScalaErasure),
           specialConstructorErasure(clazz, restpe))
-      case TypeRef(pre, `clazz`, args) =>
```
Comment (paulp) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:221]: You accidentally scooped up ArrayClass in this case.

Comment (odersky) [inline src/reflect/scala/reflect/internal/transform/Erasure.scala:221]: Good point. Will fix immediately.

**Pull request closed without merging**